### ISSUE TITLE
`initialValue`-like behaviour for the FutureProvider

### ISSUE THREAD

**Issue by RepliedSage11** [rrousselGit/riverpod#3695]
**Describe what scenario you think is uncovered by the existing examples/articles**
Counterpart to a property like `initialValue` from `pkg:provider`'s `FutureProvider` which allows the provider to have a "placeholder" value while the build methods returns for the first time. From what we have found so far, it is either not described in the documentation or simply not possible with riverpod.

PS: if this is not a documentation problem but rather a current limitation of the riverpod package please tell so, so I can maybe repurpose this issue as a proposal rather than a documentation bug.

**Describe why existing examples/articles do not cover this case**
Switching from `provider` to `riverpod` was a breeze since they are similar in most regards and other cases are well covered in the documentation, migration guides and example apps. The only thing that we couldn't find a counterpart to was something that would allow us to have an initial value for a FutureProvider that is available before the provider's `build()` method completes for the first time.

```dart
FutureProvider<int>(
  initialValue: 21,
  create: (context) => Future.value(42),
  child: MyApp(),
)
```

Or consider another example that accomplished the same task. In this (heavily simplified) case we use a ChangeNotifier which sets a "placeholder" value before fetching the state. The persisted value is immediately available to the listeners while the fresh value is fetched. Then if we want to invalidate the state, it only re-fetches the value without first reassigning in back to persisted value.
```dart
class MyChangeNotifier extends ChangeNotifier {
  MyChangeNotifier() {
    _init();
  }

  String? state;

  Future<void> _init() async {
    state = SharedPreferences.getString('key');
    notifyListeners();
    state = await fetchState();
    notifyListeners();
  }

  Future<void> invalidate() async {
    state = await fetchState();
    notifyListeners();
  }
}

```

**Additional context**
I tried asking for assistance in the Riverpod discord server and so far no-one could provide a Riverpod-native solution to this problem.

The only solution that "kinda" works is just basically copying the `_init()` method from my `ChangeNotifier` example as is. 
```dart
@riverpod
class Example extends _$Example {
  @override
  Future<String?> build() async {
    // Get the value from the cache to show it immediately
    state = SharedPreferences.get('key');

    // get fresh settings from the server
    return await fetchState();
  }
}
```
This, again, only partially accomplishes the task, since in the `ChangeNotifier` example invalidating the state would not set it back to the initial value before fetching the fresh value again. 

It also brings up another (maybe a separate issue) point which we could not get a clear understanding on from the documentation: is the `state`variable safe to directly assign in the `build()` method?

Thank you for this (and other) awesome packages. Your work is much appreciated and I am happy to further assist with this issue.

**Comment by rrousselGit** [comment]
> The only solution that "kinda" works is just basically copying the `_init()` method from my `ChangeNotifier` example as is.
> 
> ```dart
> @riverpod
> class Example extends _$Example {
>   @override
>   Future<String?> build() async {
>     // Get the value from the cache to show it immediately
>     state = SharedPreferences.get('key');
> ```

You can do:

```dart
Future<String?> build() async {
   if (!state.hasData && !state.hasError) state = <initial value>;
}
```

The initialisation should be performed only once.

**Comment by RepliedSage11** [comment]
What class should I make an extension on if I want to make a helper method for it? I can't seem to find a class that lets me access the state property and the hasData and hasError getters. 

Something like:
```dart
extension InitialValue<T> on  ... {
  void setInitialValue(T Function() builder) {
    if (!state.hasValue && !state.hasError) {
      return state = AsyncValue.data(builder());
    }
  }
}
```

**Comment by rrousselGit** [comment]
`Ref<AsyncValue<T>>` should be good enough

**Comment by RepliedSage11** [comment]
I don't think Ref has access to state since state is a property on NotifierBase class. It is also marked as `@protected`.

**Comment by rrousselGit** [comment]
It does.

**Comment by RepliedSage11** [comment]
Sorry if I am missing something obvious, but I am unable to access the state property from the Ref.

<img width="784" alt="Screenshot 2024-08-16 at 15 36 37" src="https://github.com/user-attachments/assets/94cff6f6-b126-460f-95d5-e8cc35e91959">

**Comment by rrousselGit** [comment]
Oh, seems like you need FutureProviderRef.

**Comment by RepliedSage11** [comment]
Hi, with the new 2.6.0 release Ref subclasses got deprecated. What would you suggest to use now to keep the same functionality?

**Comment by rrousselGit** [comment]
Use AsyncNotifier. It's a FutureProvider with state property.
